### DISCUSSION TITLE
Alpine.$persist is not defined before alpine:init listener execution

### DISCUSSION THREAD

**Discussion by ethan-ho** [alpinejs/alpine#2465] [7. Bugs (archived)] [answered]
I am trying to use Alpine.$persist global in a store following [documentation](https://alpinejs.dev/plugins/persist#using-alpine-persist-global) but getting `TypeError: Alpine.$persist is not a function` when creating the store.  Apparently `Alpine.$persist` hasn't been defined yet. 

Alpine.js Library and persist plugin are loaded via CDN, please following this URL for demo:  https://jsbin.com/nubonejoda/1/edit?html,console,output

**Comment by SimoTod** [comment, marked as answer]
It's a timing issue: both your custom script and the plugin use alpine:init to ensure Alpine exist. Your custom script must run after the plugin or it won't be able to find `$persist`
Your script is inlined (so it runs when the browser find it) while the plugin is deferred (so it runs when the whole page has been parsed).
You can either move your store in a js and make it a deferred script or remove defer from the plugin script.

**Reply by ethan-ho** [reply]
Thank you for the suggestion. Just a quick question, does that mean only option I have is to either:

* remove `defer` for both persist plugin and library, or
* defer declaration of `Alpine.store()` as well as all depending `Alpine.data()` declaration?

Sorry I am still very new to Alpine.js so unsure if above options may or may not be feasible.

**Reply by SimoTod** [reply]
I don't know if they are the only options but they'll definitely work.

**Reply by ethan-ho** [reply]
I've tried both ways:
* removing defer from plugin script works: https://jsbin.com/kazohiyoho/2/edit?html,console,output
* for making script tags defer, I need to make sure scripts are loaded in this order:
  ```
  <script defer src="https://unpkg.com/@alpinejs/persist@3.7.0/dist/cdn.min.js"></script>
  <script defer src="my-app.js"></script>
  <script defer src="https://unpkg.com/alpinejs@3.7.0/dist/cdn.min.js"></script>
  ```
  which works - If I was loading alpine.js before my-app.js then Alpine will complain state data aren't ready yet - but not ideal as it's not always possible to control script loading order between own JS library and external JS.
  
I guess it will make our life easier if Alpine.js can introduce an extra event lifecycle (e.g. alpine:init-plugin) for bootstrapping plugins and fired before alpine:init event, then we don't have to concern about plugins readiness.

**Reply by SimoTod** [reply]
You actually have one already (although it happened by chance): https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/lifecycle.js#L12

Use the latter for your store

**Reply by SimoTod** [reply]
The plugin event is probably a good shout though. You can try to send a PR and see what Caleb thinks if you like. (in that case, make sure you update all the existing plugins and documentation).

**Reply by ethan-ho** [reply]
I will give it a shot!

**Reply by MartijnHarmenzon** [reply]
Hello all. Is there any update on this? I am encountering a similar issue. Would be great to have.

**Reply by CamilleScholtz** [reply]
Any update on this?

**Reply by SimoTod** [reply]
There was a pull request at some point but it wasn't accepted: https://github.com/alpinejs/alpine/pull/2550

However, in alpine v3, there is already an event that fires after `alpine:init` but before Alpine tries to initialise the components: https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/lifecycle.js#L17

You can use that event for your custom code to make sure persist is loaded without caring about the order.

**Reply by ekwoka** [reply]
yes, you can use `alpine:initializing` to run on the second.

These should probably be documented. It's helpful to have all plugins use `alpine:init` and then only use `alpine:initializing` if people using the cdn have race conditions.

Ideally you don't use the cdn at all, but if you must.

**Reply by CamilleScholtz** [reply]
Thanks for the replies.

>Ideally you don't use the cdn at all, but if you must.

A little bit off-topic, but the documentation mostly shows off CDN usage. Are modules preferred? In what way are they better?

**Reply by ekwoka** [reply]
You get your own transpilation target, you don't have to deal with these race conditions, and your site doesn't need to make a connection to another origin with dns lookup and everything involved there.

You get improved compression on larger files as well.
